enhancements/api-review/custom-tags-aws.md

```diff
@@ -31,35 +37,51 @@ have the ability to apply user defined tags to many resources created by OpenShi
+- The administrator or service (in the case of Managed OpenShift) installing OpenShift can pass an arbitrary
    list of user-defined tags to the OpenShift Installer, and everything created by the installer and all other
    bootstrapped components will apply those tags to all resources created in AWS, for the life of the cluster, and where supported by AWS.
-- tags must be applied at creation time, in an atomic operation. It isn't acceptable to create an object and,
-   after some period of time, apply the tags post-creation.
+- User-defined tags can be applied at creation time, as an atomic operation.
+- User-defined tags can be updated post creation.
 
 ### Non-Goals
 
-- to reduce initial scope, tags are applied only at creation time and not reconciled. If an administrator manually
-   changes the tags stored in the infrastructure resource, behavior is undefined. See below.
-- to reduce initial scope, we are not implementing this for clouds other than AWS. We will not take any actions
-   to prohibit that later.
+- To reduce initial scope, the proposal does not apply for clouds other than AWS. There will be no actions taken to prohibit that later.
+
+### Limitations
+
+- User-defined tags for PV volumes cannot be updated but only set during creation of volumes using `--extra-tags`. Present CSI spec does not support `UpdateVolume` to update volume metadata.
+   Restarting csi driver to initiate multiple `CreateVolume` will depend on underlying storage layer idempotency implementation for the operation. It is recommended to use an interface from CSI spec which enforces idempotency guarantees for volume metadata updates.
+   In case of AWS EC2, `CreateVolume` returns `IdempotentParameterMismatch` error when identical request is not found for the same client token.
+   Also, the CSI spec `https://github.com/container-storage-interface/spec/blob/master/spec.md` mentions grpc error code
+   `6 ALREADY_EXISTS` for `a volume corresponding to the specified volume name already exists but is incompatible with the specified capacity_range, volume_capabilities, parameters, accessibility_requirements or volume_content_source`.
+- User-defined tags cannot be updated for vpc, security groups, elb(not managed by in-cluster operators), route53, subnet resources as there is no operator managing the resources post installation.
+   User-defined tags cannot be updated to an AWS resource which is not managed by an operator. In this proposal, the changes proposed and developed will be part of openshift-* namespace. External operators are not in scope.
+- User-defined tags can be updated on the following AWS resources.
+  1. EC2 instances for master and worker nodes.
+  2. Image registry.
+  3. Ingress LB.
+  4. IAM credentials by CCO in mint mode of operation.
+- OpenShift is bound by the AWS resource that allows the fewest number of tags. An AWS S3 bucket can have at most 10 tags.
```

Comment:
@sdodson @deads2k I had a conversation w/ SD about their needs/expectations and in general how we make this an api that an admin won't be surprised by the behavior of.

one of the thoughts would be rather than having a single "tags" abstraction field that:
1) will only update tags on some resources (ones we have active operator management of)
2) has to be constrained by the tag limit of the lowest resource tag limit (10)
3) needs to compete w/ machine spec which considers itself the only owner of ec2 tags, if you provide tags in that api (though possibly this could be changed)

we should instead break this out into specific api fields for each resource type we can actually manage the tags of (and no resource field for the ones we cannot).

it would make for a bit of an ugly api (and beg the next question of: why not just manage the tags via a config resource for each operator, which we already do in some cases, rather than specify them here where it's an action at a distance thing)

Another thing we discussed(possibly independent of that api shape decision) would be whether the operators that actually manage the resources should go degraded if they can't apply the requested tags.  This would at least give some feedback to the admin that the changes they tried to assert via the api did not get applied.  Unfortunately it's "action at a distance".  It would be better if the infrastructure resource reported the errors back, but introducing that sort of two way dependency between whatever manages the infra resource object, and whatever manages the aws resources, doesn't feel great to me.

thoughts?  Fundamentally I think it's important that if we're going to provide a tag-updating api interface, it be obvious to the users of that interface what aws resources will/won't be updated and what, in general, the limitations of that api are.  Otherwise they are going to add tags and then be surprised when some resources get the tags and others don't.  Which makes me think a generic "tag your aws resources via this field" api(regardless of how it is documented) is problematic.

ideas?

Reply:
That sounds like a pretty horrible user experience. What became of the approach that we set one tag, everywhere we can, and then additional tags are managed outside of OpenShift however you prefer?

Reply:
> What became of the approach that we set one tag, everywhere we can, and then additional tags are managed outside of OpenShift however you prefer?

it's all still on the table, but the problem w/ that approach is that it's not obvious when interacting with the api, which resources will get the new/updated tag and which will not.  It requires reading the documentation, assuming the documentation is correct and exhaustive to understand if a resource falls into the category of:

1) we can fully manage/update the tags
2) we can create the tags at install time, but we can't update tags
3) we can't manage the tags at all (not even create)
4) we can manage the tags as long as you don't exceed N of them, so even though you'd normally expect this tag you added to get applied to X, it didn't because X was over its tag limit, while Y did get the tag because it's not over the limit.

a single "put your tag desires here" api field doesn't get us there, not without at least a related "status" field that tells you what you actually got.  And maintaining a status field on the infra resource will be difficult since it needs to go ask each operator that manages a resource what tags the operator has actually applied to the resources.  (infra resource becomes an aggregation over the various resource-managing operators, for resources that *have* an operator that manages them)

Reply:
ah, yeah i sort of pitched that to SD.  They weren't a huge fan (basically punting tooling to them, and i think there were some more explicit concerns about how it would be done), but I think it's technically still on the table, depending on what SD and customers come back and tell us are their actual requirements.

@jewzaam @jharrington22 can better speak to their thoughts/concerns on that approach, though.

Reply:
@jharrington22  and I talked over this and we'd like to keep tagging as a part of OCP.  Not having delete is fine initially.  At some point it's possible we care about this.  Our requirement is we can set a tag on all resources created by the installer and created by OCP after installation.  We require a single tag and it's the same on all resources.  Anything beyond that is a customer use case and I'm personally OK with not tackling those at this time.  Having a tag set on all resources will allow us to create policies limiting permissions in the customer AWS account to just the cluster created/managed resources.